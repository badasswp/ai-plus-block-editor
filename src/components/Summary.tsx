import { __ } from '@wordpress/i18n';
import { check } from '@wordpress/icons'
import { select, dispatch } from '@wordpress/data';
import { Button, TextareaControl, Icon } from '@wordpress/components';
import { useState, useEffect } from '@wordpress/element';
import apiFetch from '@wordpress/api-fetch';

import Toast from '../components/Toast';

/**
 * Summary.
 *
 * This Component returns the Summary
 * label and button.
 *
 * @since 1.1.0
 *
 * @returns {JSX.Element}
 */
const Summary = (): JSX.Element => {
  const[ summary, setSummary ] = useState( '' );
  const[ isLoading, setIsLoading ] = useState( false );
  const { editPost } = dispatch( 'core/editor' ) as any;
  const {
    getCurrentPostId,
    getEditedPostContent,
    getEditedPostAttribute
  } = select( 'core/editor' );

  const content = getEditedPostContent();

  useEffect( () => {
    setSummary( getEditedPostAttribute( 'meta' )['apbe_summary'] );
  }, [] )

  /**
   * This function fires when the user clicks
   * the `Generate` button.
   *
   * @since 1.1.0
   *
   * @returns { void }
   */
  const handleClick = async (): Promise<void> => {
    setIsLoading( true );

    const response = await apiFetch(
      {
        path: '/ai-plus-block-editor/v1/sidebar',
        method: 'POST',
        data: {
          id: getCurrentPostId(),
          text: content.text || content,
          feature: 'summary'
        },
      }
    );

    const { data: aiSummary } = response as any;

    let limit = 1;

    const showAnimatedAiText = setInterval( () => {
      if ( aiSummary.length === limit ) {
        clearInterval( showAnimatedAiText );
      }
      setSummary( aiSummary.substring( 0, limit ) );
      limit++;
    }, 5 )

    setIsLoading( false );
  }

  /**
   * This function fires when the user selects
   * the AI generated result.
   *
   * @since 1.1.0
   *
   * @returns { void }
   */
  const handleSelection = (): void => {
    editPost( { excerpt: summary } );
    editPost( { meta: { apbe_summary: summary } } );
  }

  return (
    <>
      <p><strong>{ __( 'Summary', 'ai-plus-block-editor' ) }</strong></p>
      <TextareaControl
        rows={ 4 }
        value={ summary }
        onChange={ text => setSummary( text ) }
        __nextHasNoMarginBottom
      />
      <div className="apbe-button-group">
        <Button
          variant="primary"
          onClick={ handleClick }
        >
          { __( 'Generate', 'ai-plus-block-editor' ) }
        </Button>
        <Button
          variant="secondary"
          onClick={ handleSelection }
        >
          <Icon icon={ check } />
        </Button>
      </div>
      <Toast
        message={ __( 'AI is generating text, please hold on for a bit...' ) }
        isLoading={ isLoading }
      />
    </>
  )
}

export default Summary;
