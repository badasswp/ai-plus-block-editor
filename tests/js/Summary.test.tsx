import '@testing-library/jest-dom';
import { act, render, fireEvent, waitFor } from '@testing-library/react';

import apiFetch from '@wordpress/api-fetch';
import { useSelect, useDispatch } from '@wordpress/data';

import Summary from '../../src/components/Summary';

jest.mock( '@wordpress/i18n', () => ( {
	__: jest.fn( ( arg ) => arg ),
} ) );

jest.mock( '@wordpress/data', () => ( {
	useSelect: jest.fn(),
	useDispatch: jest.fn(),
} ) );

jest.mock( '@wordpress/notices', () => ( {
	store: 'core/notices',
} ) );

jest.mock( '@wordpress/components', () => ( {
	Button: jest.fn( ( { children, variant, onClick } ) => {
		return (
			<>
				<button className={ variant } onClick={ onClick }>
					{ children }
				</button>
			</>
		);
	} ),

	Icon: jest.fn( () => {
		return <>Icon</>;
	} ),

	TextareaControl: jest.fn(
		( { rows, value, onChange, __nextHasNoMarginBottom = true } ) => {
			return (
				__nextHasNoMarginBottom && (
					<>
						<textarea
							rows={ rows }
							onChange={ onChange }
							value={ value }
						/>
					</>
				)
			);
		}
	),
} ) );

jest.mock( '@wordpress/api-fetch', () => jest.fn() );

describe( 'Summary', () => {
	beforeEach( () => {
		( useSelect as jest.Mock ).mockReturnValue( {
			postId: 1,
			postContent: 'Hello World',
			postSummary: 'AI generated summary...',
			notices: [],
		} );

		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: jest.fn(),
			savePost: jest.fn(),
			removeNotice: jest.fn(),
			createErrorNotice: jest.fn(),
		} );
	} );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	it( 'renders Summary component', () => {
		const { container, getByText, getByRole } = render( <Summary /> );

		expect( container ).toMatchSnapshot();

		expect( getByText( 'AI generated summary...' ) ).toBeInTheDocument();
		expect( getByText( 'Summary' ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Icon' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Icon' } ) ).toHaveClass(
			'secondary'
		);
		expect( getByRole( 'button', { name: 'Generate' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Generate' } ) ).toHaveClass(
			'primary'
		);
	} );

	it( 'renders fetched API data from AI LLM', async () => {
		const mockEditPost = jest.fn();
		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: mockEditPost,
		} );

		( apiFetch as unknown as jest.Mock ).mockImplementation(
			jest.fn( () =>
				Promise.resolve( 'It is such a wonderful world, we live in...' )
			)
		);

		const { getByText, getByRole } = render( <Summary /> );

		expect( getByText( 'AI generated summary...' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 2 );
			expect(
				getByText( 'It is such a wonderful world, we live in...' )
			).toBeInTheDocument();
		} );
	} );

	it( 'renders error notice on API fail', async () => {
		const mockCreateErrorNotice = jest.fn();
		( useDispatch as jest.Mock ).mockReturnValue( {
			createErrorNotice: mockCreateErrorNotice,
		} );

		( apiFetch as unknown as jest.Mock ).mockRejectedValueOnce(
			new Error( 'AI LLM down...' )
		);

		const { getByText, getByRole } = render( <Summary /> );

		expect( getByText( 'AI generated summary...' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockCreateErrorNotice ).toHaveBeenCalledTimes( 1 );
			expect(
				getByText( 'AI generated summary...' )
			).toBeInTheDocument();
		} );
	} );

	it( 'saves the selected AI Summary', async () => {
		const mockEditPost = jest.fn();
		const mockSavePost = jest.fn();
		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: mockEditPost,
			savePost: mockSavePost,
		} );

		( apiFetch as unknown as jest.Mock ).mockImplementation(
			jest.fn( () =>
				Promise.resolve( 'It is such a wonderful world, we live in...' )
			)
		);

		const { getByText, getByRole } = render( <Summary /> );

		expect( getByText( 'AI generated summary...' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect(
				getByText( 'It is such a wonderful world, we live in...' )
			).toBeInTheDocument();
		} );

		const icon = getByRole( 'button', { name: 'Icon' } );
		await act( async () => {
			fireEvent.click( icon );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 4 );
			expect( mockSavePost ).toHaveBeenCalledTimes( 1 );
		} );
	} );
} );
