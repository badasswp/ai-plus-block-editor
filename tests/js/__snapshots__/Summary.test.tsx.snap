// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary renders Summary component 1`] = `
<div>
  <p>
    <strong>
      Summary
    </strong>
  </p>
  <textarea
    rows="4"
  >
    AI generated summary...
  </textarea>
  <div
    class="apbe-button-group"
  >
    <button
      class="primary"
    >
      Generate
    </button>
    <button
      class="secondary"
    >
      Icon
    </button>
  </div>
</div>
`;
